torch/_dynamo/guards.py

```diff
@@ -511,6 +577,90 @@ def combine_scopes(left, right):
+        finished_expressions = []
+
+        # A mapping of tensor_ids to tensor names
+        id_to_name_map = {}
+
+        # We should not have a shape env, or guards if we are not in config.dynamic shapes
+        # But check it anyway.
+        if not config.dynamic_shapes:
+            return finished_expressions
+
+        expr_to_tensor_ref = {}
+        guard_printer = DynamoGuardPrinter(expr_to_tensor_ref, id_to_name_map)
+
+        # tensor_check_names is the primary tensor association mechanism in dynamo.
+        # All other guards installations are driven off of it, so these ones will too.
+        for name in tensor_check_names:
+            tensor_id = tensor_check_ids[name]
+            id_to_name_map[tensor_id] = name
+
+            if tensor_id in self.output_graph.tensor_id_to_sym_shape_ref:
+                # If we made it here, this tensor_id is relevant to dynamo guard installation
+                # AND was found in the shape_env
+                tensor_ref_set = self.output_graph.tensor_id_to_sym_shape_ref[tensor_id]
+                for tensor_ref in tensor_ref_set:
+                    obj_expr = tensor_ref.expr
+                    if obj_expr in self.output_graph.shape_env.val_to_var:
+                        # Normalize in case the expr is an integer, replace with shape (ex: 10->s0)
+                        obj_expr = self.output_graph.shape_env.val_to_var[obj_expr]
+                    if obj_expr not in expr_to_tensor_ref:
+                        expr_to_tensor_ref[obj_expr] = set()
+                    expr_to_tensor_ref[obj_expr].add(tensor_ref)
+            finished_expressions.append(f"isinstance({name}, torch.Tensor)")
+        # Extract all the guard elements out of guards
+        # The guard format, atm, uses tuple position 0 for the expression
+        # and tuple position 1 for a negation. Eventually, these will be collapsed together.
+        expression_and_evaluation = [
+            (guard[0], guard[1]) for guard in self.output_graph.shape_env.guards
+        ]
+        for expression, evaluation in expression_and_evaluation:
+            expr_as_str = guard_printer.doprint(expression)
+            # We may get into a state where symbolic shape keys (all should be found in replacements)
+            # Have not been removed from the expression. This is a serious enough error state that we need to assert.
+            for key in self.output_graph.shape_env.var_to_val.keys():
+                assert str(key) not in expr_as_str, f"Unknown shape symbol {key}. "
+
+            # Certain expressions are negated in their guards.
+            if not evaluation:
+                expr_as_str = f"not {expr_as_str}"
+
+            finished_expressions.append(expr_as_str)
+
+        for expr in expr_to_tensor_ref.keys():
+            tensor_refs = expr_to_tensor_ref[expr]
+            equality_candidates = set(
+                [
+                    DynamoGuardPrinter.tensor_ref_as_str(x, id_to_name_map)
```

Comment:
each id is guaranteed to map to a unique name, right? so is it actually required to dedupe here?

Reply:
Correct